app/src/ui/changes/changes-list.tsx

```diff
@@ -560,13 +560,38 @@ export class ChangesList extends React.Component<
       { type: 'separator' },
     ]
     if (paths.length === 1) {
+      const shouldEnableContextMenu = Path.basename(path) !== GitIgnoreFileName
       items.push({
         label: __DARWIN__
           ? 'Ignore File (Add to .gitignore)'
           : 'Ignore file (add to .gitignore)',
         action: () => this.props.onIgnoreFile(path),
-        enabled: Path.basename(path) !== GitIgnoreFileName,
+        enabled: shouldEnableContextMenu,
       })
+
+      const pathComponents = path.split(Path.sep)
+      if (pathComponents.length > 0) {
+        const submenu = []
+        let assembledPath = '/'
+        for (let i = 0; i < pathComponents.length - 1; i++) {
+          assembledPath += pathComponents[i] + '/'
+          const currentPath = assembledPath
+          submenu.push({
+            label: __DARWIN__
+              ? `Ignore ${assembledPath} (Add to .gitignore)`
+              : `Ignore ${assembledPath} (add to .gitignore)`,
+            action: () => this.props.onIgnoreFile(currentPath),
+          })
+        }
```

Comment:
```suggestion
        const submenu = pathComponents.slice(0, -1).map((_, index) => {
          const label = '/' + pathComponents.slice(0, index + 1).join('/')
          return {
            label,
            action: () => this.props.onIgnoreFile(label),
          }
        })
```

Some thoughts:
1. We [prefer immutability](https://github.com/desktop/desktop/blob/development/docs/contributing/engineering-values.md#immutability-is-good) if possible so I liked your first approach of `'/' + pathComponents.slice(0, index + 1).join('/')` even tho the building of path is quite straightforward. This also naturally means you don't have to do the "currentPath = assembledPath" for the sake of making a new string object for the `onIgnoreFile` method.
2. We have decided the redundant "Ignore ... (Add to .gitignore)" is unnecessary and could even be considered harmful from an accessibility standpoint due to more cognitive overhead and repetitive screen reader announcements.
3. Lastly, since switching back to index approach slice approach in 1, it makes it pretty easy to use `map` and reduce the number of lines, but, this is more a stylistic preference. Thus, take what you like of my suggestion as long as 1 and 2 are met. :)